File: flask_app/app/models/order_item.py
from app.models.base import BaseModel, db

class OrderItem(BaseModel):
    __tablename__ = 'order_item'

    order_id = db.Column(db.String(36), nullable=False)
    product_id = db.Column(db.String(36), nullable=False)
    product_name = db.Column(db.String(100), nullable=False)
    quantity = db.Column(db.Integer, nullable=False)
    unit_price = db.Column(db.Float, nullable=False)
    total_price = db.Column(db.Float, nullable=False)
    discount_applied = db.Column(db.Boolean, default=False, nullable=False)
    discount_value = db.Column(db.Float, default=0.0, nullable=False)
    discount_id = db.Column(db.String(36), nullable=False)
    total_discount_value = db.Column(db.Float, default=0.0, nullable=False)

    def __repr__(self):
        return f"<OrderItem {self.id} - Order {self.order_id}>"

    @property
    def original_total(self):
        return round(self.unit_price * self.quantity, 2)

    @property
    def effective_total(self):
        discount = self.discount_value if self.discount_applied else 0.0
        return round(self.original_total - discount, 2)

File: springboot_api/src/main/java/com/example/membership_api/model/OrderItem.java
package com.example.membership_api.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.EqualsAndHashCode;


@Data
@Entity
@Table(name = "order_item")
@EqualsAndHashCode(callSuper = true)
public class OrderItem extends  BaseModel{

    @Column(name = "order_id", nullable = false)
    private String orderId;

    @Column(name = "product_id", nullable = false)
    private String productId;

    @Column(name = "product_name", nullable = false, length = 100)
    private String productName;

    @Column(nullable = false)
    private Integer quantity;

    @Column(name = "unit_price", nullable = false)
    private Double unitPrice;

    @Column(name = "total_price", nullable = false)
    private Double totalPrice;

    @Column(name = "discount_applied", nullable = false)
    private Boolean discountApplied = false;

    @Column(name = "discount_value", nullable = false)
    private Double discountValue = 0.0;

    @Column(name = "discount_id", nullable = true)
    private String discountId;

    @Column(name = "total_discount_value", nullable = false)
    private Double totalDiscountValue = 0.0;

    public Double getOriginalTotal() {
        return Math.round(unitPrice * quantity * 100.0) / 100.0;
    }

    public Double getEffectiveTotal() {
        double discount = (discountApplied && discountValue != null) ? discountValue : 0.0;
        return Math.round((getOriginalTotal() - discount) * 100.0) / 100.0;
    }
}
